prow/config/agent.go

```diff
@@ -40,70 +47,217 @@ type Agent struct {
+		// for files in a directory trigger events for the directory
+		if info != nil && info.IsDir() {
+			if isCM, err := IsConfigMapMount(path); err != nil {
+				return fmt.Errorf("Failed to check is path %s is configmap mounted: %v", path, err)
+			} else if isCM {
+				cms.Insert(path)
+				// configmaps can't have nested directories
+				return filepath.SkipDir
+			} else {
+				dirs.Insert(path)
+				return nil
+			}
+		}
+		return nil
+	})
+	return cms, dirs, err
+}
+
+func watchConfigs(ca *Agent, prowConfig, jobConfig string, additionals ...func(*Config) error) error {
+	cmEventFunc := func() error {
+		c, err := Load(prowConfig, jobConfig, additionals...)
+		if err != nil {
+			return err
+		}
+		ca.Set(c)
+		return nil
+	}
+	// We may need to add more directories to be watched
+	dirsEventFunc := func(w *fsnotify.Watcher) error {
+		c, err := Load(prowConfig, jobConfig, additionals...)
+		if err != nil {
+			return err
+		}
+		ca.Set(c)
+		// TODO(AlexNPavel): Is there a chance that a ConfigMap mounted directory may appear without making a new pod? If yes, handle that.
+		_, dirs, err := ListCMsAndDirs(jobConfig)
+		if err != nil {
+			return err
+		}
+		for dir := range dirs {
+			// Adding a file or directory that already exists in fsnotify is a no-op, so it is safe to always run Add
+			if err := w.Add(dir); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	errFunc := func(err error, msg string) {
+		logrus.WithField("prowConfig", prowConfig).
+			WithField("jobConfig", jobConfig).
+			WithError(err).Error(msg)
+	}
+	cms := sets.NewString()
+	dirs := sets.NewString()
+	// TODO(AlexNPavel): allow empty jobConfig till fully migrate config to subdirs
 	if jobConfig != "" {
-		jobConfigStat, err := os.Stat(jobConfig)
+		stat, err := os.Stat(jobConfig)
 		if err != nil {
-			logrus.WithField("jobConfig", jobConfig).WithError(err).Error("Error loading job configs.")
-			return time.Time{}, err
+			return err
 		}
-
-		if jobConfigStat.ModTime().After(recentModTime) {
-			recentModTime = jobConfigStat.ModTime()
+		// TODO(AlexNPavel): allow single file jobConfig till fully migrate config to subdirs
+		if stat.IsDir() {
+			var err error
+			// jobConfig points to directories of configs that may be nested
+			cms, dirs, err = ListCMsAndDirs(jobConfig)
+			if err != nil {
+				return err
+			}
+		} else {
+			// If jobConfig is a single file, we handle it identically to how prowConfig is handled
+			if jobIsCMMounted, err := IsConfigMapMount(filepath.Dir(jobConfig)); err != nil {
+				return err
+			} else if jobIsCMMounted {
+				cms.Insert(filepath.Dir(jobConfig))
+			} else {
+				dirs.Insert(jobConfig)
+			}
 		}
 	}
-	return recentModTime, nil
+	// The prow config is always a single file
+	if prowIsCMMounted, err := IsConfigMapMount(filepath.Dir(prowConfig)); err != nil {
+		return err
+	} else if prowIsCMMounted {
+		cms.Insert(filepath.Dir(prowConfig))
+	} else {
+		dirs.Insert(prowConfig)
```

Comment:
You call this `dirs` but insert files

Reply:
Hmm. I have the `jobConfig` dirs added as the first items here, `prowConfig` would be the only item that's a file instead. Would `files` be a better name overall?

Reply:
Lol maybe `inodes` ... it's fine as is